### DISCUSSION TITLE
Datasette for Dashboards

### DISCUSSION THREAD

**Discussion by jcmkk3** [simonw/datasette#960] [General]
A couple of days ago Simon was posting a [question on twitter](https://twitter.com/simonw/status/1302427504477569025) about web component frameworks and referenced that he was looking at a use case for creating dashboards. I just wanted to chime in to say that I'm really excited to see what direction he takes this. I have always felt that this would be a great use for Datasette.

Simon, if you need any specific feedback about user's experience with other dashboarding tools, then I (and I'm sure other community members) would be happy to help.

**Comment by simonw** [comment]
I would absolutely love your input on this!

My idea at the moment is extremely vague. I'm interested in using Web Components for this, for a few reasons:

- It's an excuse to try out a technology I've been meaning to use for a while
- I like the idea that individual dashboard components can be built using whatever technology the widget author is most familiar with
- Web Components provide CSS isolation, which means individual dashboard widgets can style themselves exactly how they like and can then be dropped onto a page
- I figure that the first few overall dashboards can be constructed as regular HTML - which can then inform the visual Dashboard-builder tools that are likely to follow

Imagine being able to construct an initial dashboard something like this:

```html
<html>
<head><title>My Dashboard</title>
<script type="module" src="default-datasette-dashboard-widgets.js"></script>
</head>
<body>
<h1>My Dashboard</h1>
<h2>Recent stars</h2>
<table-summary
  url="https://github-to-sqlite.dogsheep.net/github/stars"
  size="5"
></table-summary>
<h2>Commits per day</h2>
<bar-chart
  database="https://github-to-sqlite.dogsheep.net/github"
  sql="select substr(author_date, 0, 11) as day, count(*) as commit_count from commits where author = 9599 group by day order by day desc limit 10"
  x="day"
  y="commit_count"
></bar-chart>
<h2>Covid worst counties</h2>
<table-summary
  url="https://covid-19.datasettes.com/covid/latest_ny_times_counties_with_populations?_sort_desc=deaths_per_million"
  size="10"
></table-summary>
</body>
</html>
```

The `<table-sumary>` elements would output a table. The `<bar-chart>` element would output essentially just the bar chart from this page: https://github-to-sqlite.dogsheep.net/github?sql=select+substr(author_date%2C+0%2C+11)+as+day%2C+count(*)+as+commit_count+from+commits+where+author+%3D+9599+group+by+day+order+by+day+desc+limit+10&p0=9599#g.mark=bar&g.x_column=day&g.x_type=ordinal&g.y_column=commit_count&g.y_type=quantitative

**Reply by koaning** [reply]
Interesting! I was contemplating something similar but I hadn't considered adding sql to the html. That might make things much easier. I don't know if you'd *really* need webcomponents for this though.

**Comment by jcmkk3** [comment]
This is a bit of a stream of thought, but I'm really just trying to think about how Datasette would compare, how I imagine it working, and what makes me excited about using it for this purpose.

## Datasette's Foundation
I think that Datasette already has some major advantages over existing dashboard solutions.

1. It directly exposes SQL to the user, which means that their ability to manipulate data is not limited by whatever abstraction is added by the BI tool vendor.
2. It has best-in-class data export capabilities. This includes the ability for the user to customize the the subset of data for their needs. Also having a direct link with all of the query parameters to perform the same export in the future.
3. It has rich tabular representations of the data. Outside of Excel, most BI tools are actually really limited/bad at presenting data in a tabular format.
4. All configuration can be done through text based files so all of the advantages of automation, version control, CI/CD, etc. apply.

## Layout and Styling
I think most dashboarding tools end up being limited or leaky abstractions over web technologies already, so just exposing those technologies to the users will end up empowering them and providing them with more transferable skills (like Datasette already does with SQL). The biggest obstacle for analysts to build dashboards with these technologies has been the back end which is what Datasette is providing (and more).

## Visualizations
JavaScript already has the best in class visualization libraries (D3, Vega/Vega-Lite, Plotly). Just opening up this world to a larger user base will greatly expand what they can create. Even some of the tools with better visualization capabilities like Tableau still can be easily out classed by these existing libraries.

## Controls and Interactions
Many dashboarding tools focus on the visualization elements and maybe some sort of image type of element. These two types of elements are abused to create buttons, sidebars, navigational menus, and all kinds of other hacks. All of this to mimic the navigational style that users have become familiar with in web based tools.

One type of control that is typically built-in, however, is tabs. It is very common for a dashboard to have multiple tabs with different views/subjects analyzed on each tab. This is really just a navigational element between pages, but is presented as tabs.

Interactions linking separate visuals and also linking visuals to controls is something that has varing support in BI tools. Tableau probably offers the richest interactions, but there is a lot of potential for much better in a web based dashboard. This is what seems most tricky to me when thinking about building a dashboard based on Datasette. I'm guessing that this would be accomplished with named parameters in the SQL queries that are tied to selections from the charts or controls somehow.

**Comment by jcmkk3** [comment]
What I see a lot in other dashboard frameworks like Dash, Panel, Streamlit, Voila, and Shiny is that they come with a lot of pre-styled components. I think that this can be useful for some audiences, but I would really like to see an option that is less opinionated on style that makes it easy for the designer to make it their own.

If you look at some of the [top dashboards built in Tableau](https://vizification.com/2020/01/2019-top-500-favorites), for instance, you will see a lot of eclectic styles. These don't all just look like admin dashboards. Of course, in corporations, there are plenty of less inspired designs.

I mentioned before about how Datasette takes care of the backend issue. The other place that seems particularly tricky to an analyst like myself is wiring up the visuals and controls to each other and the data. This is a place where maybe the web components could help. I don't feel proficient enough in web development to have a clear idea about how this would work, but this is where the other frameworks above assist the designer/developer.

**Comment by rclement** [comment]
Sorry to debunk this old discussion, but recently I have had the need for a "baked-data" BI tool for several use-cases: a mix between Datasette and Metabase if you will.

FYI, I successfully implemented the "baked-data" pattern by:

1. Running ETL jobs in scheduled CI/CD pipelines
2. Embedding an SQLite database into a Metabase Docker image
3. Deploy the image to Docker-based serverless platforms (Heroku, CloudRun, etc.).

It works really well for my use-cases. However, Metabase is heavy (like really heavy, thanks JVM) and always requires authentication as internal BI tool is the main target. Also, I like the ecosystem and spirit of the Datasette project better, especially towards publishing Open Data.

Is there currently any work/discussion/thinking about building highly composable dashboards and data visualizations with Datasette? (either built-in or as a plugin)

**Reply by jcmkk3** [reply]
You may actually have perfect timing. Just this morning, I saw that @koaning (of [https://calmcode.io/](https://calmcode.io/) fame) started a new repo for [subsette](https://github.com/koaning/subsette), a project to help with "dash-boarding in Datasette". I don't know what he has planned, but judging by his understanding about how parts of an ecosystem can come together and his other projects, I imagine that it will be a great addition.

**Reply by koaning** [reply]
Hi all. 

So ... I started this project mere hours ago so I wouldn't want to claim any progress yet. Also, how did I miss this discussion?!

Having said that, I am exploring doing dashboards on top of datasette. I've played around with metabase and although I do like it ... I'm looking for something a bit less "clicky click" and a bit more "hacky hacky".

Here's a screenshot of a version 0 of an editor on top of datasette that allows you to fully customise a vega chart. 

![image](https://user-images.githubusercontent.com/1019791/114231479-cda2f200-997a-11eb-9587-08d9ab65f2cb.png)

The idea is that this vega meta-data can then be copied/pasted in html. I've written a hacky library called [justcharts](https://github.com/koaning/justcharts) that allows you to declare a chart in pure html via something like:

```jsx
<vegachart style="width: 50%">
    {
        "data":{
            "url": "/bigmac.csv?sql=select+rowid%2C+date%2C+currency_code%2C+name%2C+local_price%2C+dollar_ex%2C+dollar_price+from+bigmac+where+currency_code+in+%28%22GBP%22%2C+%22EUR%22%2C+%22CAD%22%29+order+by+rowid+limit+101&_size=max",
            "format": {
                "type": "csv"
            }
        },
        "mark": "line",
        "encoding": {
            "x": {"field": "date", "type": "temporal"},
            "y": {"field": "local_price", "type": "quantitative"},
            "color": {"field": "currency_code", "type": "nominal"}
        },
        "width": "container"
    }
</vegachart>
```

This means you can make a simple html file and get a result like: 

![image](https://user-images.githubusercontent.com/1019791/114231789-2f635c00-997b-11eb-9323-15bb07b429ca.png)

I've literally *just* started working on this. I'm still very much in the "am I even solving it the right way"-mode. But this is the direction I'm currently exploring.

I'm especially looking forward to pushing to github, deploying a datasette instance with authentication and being able to point my iphone at "my personal dashboard".

**Reply by koaning** [reply]
I am now contemplating if something like this might work. You sometimes want to link elements.

```html
<select sql="select distinct country_code from bigmac" id="varname"/>
<vegachart sql="select * from bigmac where country_code in countries" countries="varname">
   {...}
</vegachart>
```

Then again, there's a 80/20 rule here. The linking would make for a sweet demo, but in real life I'm probably be more interested in just a few simple line/scatter/bar charts.

Note to self, vegalite also [supports selectors](https://vega.github.io/vega-lite/docs/bind.html#input-element-binding) in their json schema.

**Reply by rclement** [reply]
Thank you @koaning for sharing your train of thoughts and first experiments!

> Having said that, I am exploring doing dashboards on top of datasette. I've played around with metabase and although I do like it ... I'm looking for something a bit less "clicky click" and a bit more "hacky hacky".

This is definitely one major painpoint of Metabase: one cannot configure data-sources/visualizations/dashboards using payload configs (e.g. YAML).

> I'm especially looking forward to pushing to github, deploying a datasette instance with authentication and being able to point my iphone at "my personal dashboard".

Mmmmh I see... From my perspective, I like the monolithic approach and ease of deployment of the "backed-data" pattern.
I think the static and serverless bundle fills a niche use-case a lot of people are always plumbing with quick and dirty hacks around Seaborn, Matplotlib, Plotly, D3, and the like. In the same vein as SSG for editorial content.

Anyway, I'll think about all of these great ideas over the week-end and bring some more feedback if any!

**Reply by rclement** [reply]
@koaning you can discard my last observations about the "monolithic approach": I went through your experiment and saw that you are only defining custom templates and pages for Datasette (and not building an external tool as I first thought).

I experimented quickly this week-end and I think there is some potential about building a nice configurable dashboarding/reporting tool on top of Datasette, using custom templates & pages. I planned a discussion with Simon to talk about it next week, will report if any new ideas come out of this!

**Reply by koaning** [reply]
The reason I'm building on top of datasette for now is because of CORS. The [vega-editor](https://vega.github.io/) comes with a node backend that can make requests on my behalf but if it all came via my front-end then backends are likely to block the request. That said, the ideas in `justcharts` should be generally applicable. 

<details>
  <summary><b>Click me for a cool related demo.</b></summary>

I actually got this work *just now*. I'm not just doing git scraping. I'm also doing the data hosting there. That means that github will kollekt *and host*. 

The github repo is [here](https://github.com/koaning/kolektor) the dashboard from github pages is [here](https://koaning.github.io/kolektor/index.html). It's a demo using the `justcharts` trick, but honestly most of the heavy lifting is done by vega. 

</details>

**Reply by koaning** [reply]
I just released the first version of `justcharts`. It's hacky, but it's exactly what I need now for static dashboards.

**Comment by jcmkk3** [comment]
I really enjoy brainstorming all of the possibilities of Datasette as a BI tool. What I've been thinking about recently is the overall workflow of analyzing data then turning it into some sort or artifact to share. What I'd like is the ability to perform a literate analysis and then use some subset of that to turn into a dashboard, slide deck, or document.

The way that I imagine it working is by writing in a markdown document within the Datasette instance. I could use code blocks for sql queries, then reference them in other code blocks that contain vega-lite specs. The results of this may be all that I need. I could either share a link to the HTML rendered or export it to some document format (pdf, doc, odf).

If I needed more, I would be given the option to create a web page (dashboard). This would open up a view like CodePen or Glitch. I'd be able to reference the tables and visuals that I created in my exploratory document and add their `id` to a web component in the HTML.

There are many projects that provide a way do generate slides with markdown or html so I think that use-case could be easily fulfilled.

While this would not be nearly as easy as a pure GUI tool or as flexible as Observable and/or a custom website, I think that it would hit a sweet spot between them. The skills learned in this setup (sql, markdown, json, html, css) would be transferrable in the future. I think that for many uses, a small subset of knowledge is all that would be needed and the tool could help with autocompletion and generating some code/skeletons.

Some Vega-Lite links
- [vegadown](https://vegawidget.github.io/vegadown/) - An example of vega-lite embeded in markdown
- [deneb](https://coacervo.co/deneb_public_preview) - An example of a tool generating some vega-lite skeletons
- [vega-lite editor](https://vega.github.io/editor/#/examples/vega-lite/bar) - The official editor. The auto-completion really adds to the authoring experience.

Some future ideas
- Allow sql code blocks to refer to other sql blocks as sources. This would allow the user to build up complex queries in pieces and see intermediate results.
- Related to the above idea, maybe push down go [dbt](https://github.com/fishtown-analytics/dbt) ([sqlite plug-in](https://github.com/codeforkjeff/dbt-sqlite)) as the query runner. That would provide the benefit of [extending the sql](https://docs.getdbt.com/docs/building-a-dbt-project/jinja-macros) through jinja.
- Speaking of jinja, since datasette already pulls it in, it could be enabled in the html editor for building dashboards.
- Query push-down in vega-lite so that the implicit and explicit transformations happen in the sql layer intead of client side javascript. There are multiple projects exploring this space like [this one](https://github.com/omnisci/vega-transform-omnisci-core) for OmniSci.
- Maybe an [Arrow](https://arrow.apache.org/) export for Datasette to reduce the size of data being transferred?
- Other charting libraries like D3 or Plotly
- Some ready to go html widgets like the ones from [Observable](https://github.com/observablehq/inputs). Vega-lite already has some built-in, of course.

**Reply by rclement** [reply]
Thanks @jcmkk3 for your thoughts!
I think most of your desires can already be achieved using a Jupyter / Observable Notebook against a Datasette instance.

However, while doing some more coding on [`datasette-dashboards`](https://github.com/rclement/datasette-dashboards) this week-end, I realized something:

- I plan on adding declarative custom layouts for dashboards
- I plan on adding Markdown elements as chart type (shamelessly copied from Metabase)
- Combining custom layouts and Markdown text would allow for "literate notebooks" (see [Markdown in Metabase Dashboards](https://www.metabase.com/learn/building-analytics/dashboards/markdown))

The current state of this plugin is basic but gets the job done if you want to share a dashboard with a bunch of charts, defined only using a single metadata config file (you can literally slap Vega-Lite spec in charts and it just works). I updated the [demo](https://datasette-dashboards-demo.vercel.app/-/dashboards/job_offers_stats) and see what we can already achieve with [a single metadata file](https://github.com/rclement/datasette-dashboards/blob/master/demo/metadata.yml) without much boilerplate (e.g. choropleth maps, Metabase-like interactivity).

If this work resonates with what you are hoping for, please come share you valuable input and contribute! (I started a [meta issue](https://github.com/rclement/datasette-dashboards/issues/1) with all the improvement ideas I had so far)

**Reply by rclement** [reply]
@jcmkk3 I just released `datasette-dashboards` version 0.1.2 which includes:

- Markdown components
- Custom layout

I updated the online demo to showcase a dashboard with an analysis note next to charts, opening the door for "storytelling" analysis.

![Datasette Dashboards Demo](https://raw.githubusercontent.com/rclement/datasette-dashboards/master/demo/datasette-dashboards-demo.png)

**Reply by jcmkk3** [reply]
This looks great! It really fits in pretty seamlessly with the styling/navigation of Datasette.

**Comment by rclement** [comment]
Here is my first experiment defining and generating data dashboards with Datasette: https://github.com/rclement/datasette-dashboards

For now the idea is:

- Install the `datasette-dashboards` plugin (not ready yet)
- Define dashboards in `metadata.yml` / `metadata.json` using SQL queries + Vega-Lite charts terminology:
```yaml
plugins:
  datasette-dashboards:
    my_dashboard:
      title: My Dashboard
      description: Showing some nice metrics
      visualizations:
        - title: Number of events by day
          chart: line
          encoding:
            x: { field: day, type: temporal }
            y: { field: count, type: quantitative }
          db: jobs
          query: SELECT date(date) as day, count(*) as count FROM events GROUP BY day ORDER BY day
```
- Go to `/-/dashboards` to find the dashboards index

This is really really basic for now.
What am I missing the most? A visual query builder turning data into visualizations, allowing you to copy the config to the metadata.
Basically, as if Metabase would allow to define dashboards and visualization through a static configuration...

BTW, I spoke with @simonw today, he seemed to be enjoying the idea! Having more data analysis tools like these is something that could bring some less tech-savy audiences to use Datasette in day-to-day use (students, business departments, etc.).

Here are some screenshots:

<img width="1440" alt="datasette-dashboards-index" src="https://user-images.githubusercontent.com/1238873/115938349-a11ec800-a49a-11eb-96a3-ac9691b9771a.png">
<img width="1440" alt="datasette-dashboards-view" src="https://user-images.githubusercontent.com/1238873/115938352-a24ff500-a49a-11eb-8078-c96c9534c3bd.png">

**Reply by rclement** [reply]
Ok, I had quite some time on my hand today so here is my baby plugin taking some shape:

- [`datasette-dashboards`](https://pypi.org/project/datasette-dashboards/) plugin is on PyPI
- A live demo is available at [https://datasette-dashboards-demo.vercel.app](https://datasette-dashboards-demo.vercel.app/-/dashboards)

**Reply by koaning** [reply]
That's pretty neat! 

In particular, I quite like this formatter:

```yaml
      layout:
        - [chart1, chart2, chart2, chart5]
        - [chart1, chart3, chart4, chart5]
```

**Reply by rclement** [reply]
Thanks @koaning !

Yeah, the layout property is a mere adaptation of the CSS Grid template area.
I did not want to introduce some kind of `ref` field for charts to reference them, so I came up with this way of referencing them with `chart{i+1}`: I am not completely satisfied with its implicit character but it fits the bill as declarative and human-readable abstraction.
If you have a better suggestion, I'm all ears!

**Reply by koaning** [reply]
Why didn't you add a `ref`? It sounds like a neat idea in my mind. I might call it `alias` or `name` but having a user declare a reference sounds less error-prone to me.

**Reply by rclement** [reply]
I did not want to "overload" chart configuration with technical details (references). But I agree with you it is error-prone and that's why I said I am not completely satisfied with this system. `alias` might be a good fit as "human-readable" (I'm afraid `name` being too close to `title`), good call thanks!

And to be transparent, one of my use-cases for these declarative static dashboards is students: I teach the rudiments of data engineering/analysis/science to initially non tech-savy people and the less technicalities there are the better (e.g. they hate having to fiddle with plotting libraries such as `matplotlib` or `seaborn`).

**Reply by koaning** [reply]
Yeah, vegalite is definitely more like a proper grammar as opposed to merely a plotting API. I can see myself making a few on this on https://calmcode.io in a month or two! I think the `alias` would be a nice feature to add though.

**Reply by rclement** [reply]
Thanks for pointing out Calmcode, this seems an amazing resource for learners! I will definitely spread the word to my students! Are you producing the content all by yourself? Is the platform Open Source ?

**Reply by koaning** [reply]
Yes at the moment I'm the sole maintainer. The datasette course is the most popular one on the platform at the moment. There's a github where folks can post issues but it's not open source. I contemplated doing that, but there are some question/answer type things where you do not want the answer to be publicly available.

**Reply by rclement** [reply]
> there are some question/answer type things where you do not want the answer to be publicly available.

Okay fair enough. But from what I understand, this is not an "open-source" problem, this is a permission/access-control issue (RBAC), which is solved at runtime on the back-end side of things.

I was just asking if it was OSS because the platform is really slick and the content is high-quality, and if it was built in the open that would be cherry on top!

**Reply by rclement** [reply]
Back on track with the dashboards: I introduced the `alias` property to reference charts in dashboard layout
https://github.com/rclement/datasette-dashboards#usage

**Reply by koaning** [reply]
@rclement there have been moments where I thought open-sourcing it would be neat. But since it's something that's hard to reverse I figured delaying such decision would be best. I'm not looking forward to a lot of review work, but a part of me is thinking about ways people might collaborate. I've been running the project for a year now and although I really like where it's at ... I haven't really figured out what I'd like it to be in a years time.
